File: src/director_ai/research/consciousness/tcbo.py
# ─────────────────────────────────────────────────────────────────────
# Director-Class AI — TCBO (Topological Consciousness Boundary Observable)
# (C) 1998-2026 Miroslav Sotek. All rights reserved.
# Contact: www.anulum.li | protoscience@anulum.li
# ORCID: https://orcid.org/0009-0009-3560-0851
# License: GNU AGPL v3 | Commercial licensing available
# ─────────────────────────────────────────────────────────────────────
"""
Topological Consciousness Boundary Observable (TCBO).

Extracts a single scalar p_h1(t) ∈ [0, 1] from multichannel phase data via
persistent homology (H1 cycles in Vietoris-Rips complexes of delay-embedded
signals). When p_h1 > τ_h1 (default 0.72), the "consciousness gate" opens.

Pipeline:
  1. Multichannel signal → delay embedding (Takens' theorem)
  2. Sliding window → point cloud
  3. Vietoris-Rips persistent homology (H1)
  4. Max H1 persistence → logistic squash → p_h1 ∈ [0, 1]

Scientific grounding:
  - H1 persistent homology validated for brain state classification
    (Wang et al. 2025, Santoro et al. 2024)
  - Gap junctions (connexin-36) essential for gamma oscillations
  - Delay embedding (Takens' theorem) reconstructs attractor topology

Dependencies:
  - ripser (optional; pure-NumPy fallback provided)
  - numpy (required)
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np

# ── Ripser detection ─────────────────────────────────────────────────

try:
    from ripser import ripser as _ripser_fn

    _HAS_RIPSER = True
except ImportError:
    _HAS_RIPSER = False


# ── Pure-numpy fallback for H1 persistence ───────────────────────────


def _ripser_fallback(point_cloud: np.ndarray, maxdim: int = 1) -> dict:
    """Minimal fallback when ripser is unavailable.

    Estimates H1 persistence from point-cloud spread. Not real persistent
    homology — provides a plausible scalar so the pipeline can run.
    """
    centered = point_cloud - np.mean(point_cloud, axis=0)
    radial = np.linalg.norm(centered, axis=1)
    spread = float(np.std(radial)) if len(radial) > 1 else 0.0

    b0 = np.array([[0.0, np.inf]])
    if spread > 1e-12:
        b1 = np.array([[0.15 * spread, 0.85 * spread]])
    else:
        b1 = np.empty((0, 2), dtype=np.float64)

    dgms = [b0, b1]
    if maxdim >= 2:
        dgms.append(np.empty((0, 2), dtype=np.float64))
    return {"dgms": dgms}


def _compute_ripser(point_cloud: np.ndarray, maxdim: int = 1) -> dict:
    """Dispatch to real ripser or fallback."""
    if _HAS_RIPSER:
        return _ripser_fn(point_cloud, maxdim=maxdim)
    return _ripser_fallback(point_cloud, maxdim=maxdim)


# ── Delay embedding (Takens) ────────────────────────────────────────


def delay_embed(
    x: np.ndarray,
    embed_dim: int = 3,
    tau_delay: int = 1,
) -> np.ndarray:
    """Delay-coordinate embedding of a 1-D signal.

    Parameters
    ----------
    x : ndarray (T,) — univariate time series.
    embed_dim : int — embedding dimension m.
    tau_delay : int — delay in samples.

    Returns
    -------
    Z : ndarray (T_out, embed_dim) where T_out = T - (m-1)*tau.
    """
    T = len(x)
    m = embed_dim
    offset = (m - 1) * tau_delay
    if offset >= T:
        return np.empty((0, m), dtype=np.float64)

    out_T = T - offset
    Z = np.empty((out_T, m), dtype=np.float64)
    for k in range(m):
        start = offset - k * tau_delay
        Z[:, k] = x[start : start + out_T]
    return Z


def delay_embed_multi(
    X: np.ndarray,
    embed_dim: int = 3,
    tau_delay: int = 1,
) -> np.ndarray:
    """Delay-embed each channel then concatenate.

    Parameters
    ----------
    X : ndarray (T, n_channels)
    embed_dim, tau_delay : as above.

    Returns
    -------
    Z : ndarray (T_out, n_channels * embed_dim)
    """
    n_ch = X.shape[1]
    parts = [delay_embed(X[:, ch], embed_dim, tau_delay) for ch in range(n_ch)]
    if any(p.shape[0] == 0 for p in parts):
        return np.empty((0, n_ch * embed_dim), dtype=np.float64)
    return np.hstack(parts)


# ── Persistence → probability ────────────────────────────────────────


def persistence_to_probability(
    s_h1: float,
    s0: float = 0.0,
    beta: float = 8.0,
) -> float:
    """Logistic squashing of H1 persistence to [0, 1].

    p_h1 = 1 / (1 + exp(-β · (s_h1 - s0)))
    """
    arg = -beta * (s_h1 - s0)
    arg = max(-500.0, min(500.0, arg))
    return 1.0 / (1.0 + np.exp(arg))


def s0_for_threshold(
    s_tau: float = 0.5,
    p: float = 0.72,
    beta: float = 8.0,
) -> float:
    """Compute s0 so that persistence_to_probability(s_tau, s0, beta) == p."""
    p = float(np.clip(p, 1e-12, 1.0 - 1e-12))
    return s_tau - (1.0 / beta) * np.log(p / (1.0 - p))


# ── TCBOObserver ─────────────────────────────────────────────────────


@dataclass
class TCBOConfig:
    """Configuration for the TCBO observer."""

    embed_dim: int = 3
    tau_delay: int = 1
    window_size: int = 50
    tau_h1: float = 0.72
    beta: float = 8.0
    s0: float = 0.0
    persistence_threshold: float = 0.05
    subsample_max: int = 500
    compute_every_n: int = 1


class TCBOObserver:
    """Topological Consciousness Boundary Observable.

    Produces a single scalar p_h1(t) ∈ [0, 1] from multichannel phase data.

    Parameters
    ----------
    N : int — number of channels (SCPN layers).
    config : TCBOConfig — observer configuration.
    """

    def __init__(
        self,
        N: int = 16,
        config: TCBOConfig | None = None,
    ) -> None:
        self.N = N
        self.cfg = config if config is not None else TCBOConfig()

        self._buffer: list[np.ndarray] = []
        self._max_buffer = (
            self.cfg.window_size + (self.cfg.embed_dim - 1) * self.cfg.tau_delay + 10
        )

        self.p_h1: float = 0.0
        self.s_h1: float = 0.0
        self.is_conscious: bool = False
        self.h1_error: float = self.cfg.tau_h1
        self._step_count: int = 0
        self._dgms: list | None = None

    def push(self, theta: np.ndarray) -> None:
        """Push a new phase vector into the rolling buffer."""
        self._buffer.append(theta.copy())
        if len(self._buffer) > self._max_buffer:
            self._buffer = self._buffer[-self._max_buffer :]

    def compute(self, force: bool = False) -> float:
        """Compute p_h1 from the current buffer.

        Returns
        -------
        p_h1 : float in [0, 1].
        """
        self._step_count += 1

        if not force and (self._step_count % self.cfg.compute_every_n != 0):
            return self.p_h1

        min_needed = (
            self.cfg.embed_dim - 1
        ) * self.cfg.tau_delay + self.cfg.window_size
        if len(self._buffer) < min_needed:
            return self.p_h1

        signal = np.array(self._buffer[-min_needed:])
        Z = delay_embed_multi(signal, self.cfg.embed_dim, self.cfg.tau_delay)
        if Z.shape[0] == 0:
            return self.p_h1

        cloud = Z[-self.cfg.window_size :]
        if cloud.shape[0] > self.cfg.subsample_max:
            idx = np.random.choice(
                cloud.shape[0], self.cfg.subsample_max, replace=False
            )
            cloud = cloud[idx]

        result = _compute_ripser(cloud, maxdim=1)
        self._dgms = result["dgms"]

        h1 = result["dgms"][1]
        if h1.shape[0] > 0:
            finite_mask = np.isfinite(h1[:, 1])
            if np.any(finite_mask):
                lifetimes = h1[finite_mask, 1] - h1[finite_mask, 0]
                sig_mask = lifetimes > self.cfg.persistence_threshold
                if np.any(sig_mask):
                    self.s_h1 = float(np.max(lifetimes[sig_mask]))
                else:
                    self.s_h1 = 0.0
            else:
                self.s_h1 = 0.0
        else:
            self.s_h1 = 0.0

        self.p_h1 = persistence_to_probability(self.s_h1, self.cfg.s0, self.cfg.beta)
        self.is_conscious = self.p_h1 > self.cfg.tau_h1
        self.h1_error = max(0.0, self.cfg.tau_h1 - self.p_h1)

        return self.p_h1

    def push_and_compute(self, theta: np.ndarray, force: bool = False) -> float:
        """Push + compute in one call."""
        self.push(theta)
        return self.compute(force=force)

    def get_state(self) -> dict:
        """Return serialisable observer state."""
        return {
            "p_h1": round(self.p_h1, 6),
            "s_h1": round(self.s_h1, 6),
            "is_conscious": self.is_conscious,
            "h1_error": round(self.h1_error, 6),
            "tau_h1": self.cfg.tau_h1,
            "buffer_size": len(self._buffer),
            "has_ripser": _HAS_RIPSER,
        }

    def reset(self) -> None:
        """Clear buffer and reset state."""
        self._buffer.clear()
        self.p_h1 = 0.0
        self.s_h1 = 0.0
        self.is_conscious = False
        self.h1_error = self.cfg.tau_h1
        self._step_count = 0
        self._dgms = None


# ── TCBOController ───────────────────────────────────────────────────


@dataclass
class TCBOControllerConfig:
    """PI controller configuration for gap-junction coupling."""

    tau_h1: float = 0.72
    Kp: float = 0.8
    Ki: float = 0.2
    integral_min: float = -5.0
    integral_max: float = 5.0
    kappa_min: float = 0.0
    kappa_max: float = 5.0
    history_len: int = 100


class TCBOController:
    """PI controller driving gap-junction kappa from p_h1 deficit.

    Error signal: e_h1(t) = max(0, τ_h1 - p_h1(t))  (deficit-only)
    """

    def __init__(
        self,
        config: TCBOControllerConfig | None = None,
    ) -> None:
        self.cfg = config if config is not None else TCBOControllerConfig()
        self._integral: float = 0.0
        self._last_error: float = 0.0
        self._p_h1_history: list[float] = []
        self._kappa_history: list[float] = []
        self._error_history: list[float] = []

    def compute_error(self, p_h1: float) -> float:
        """Deficit-only error: positive when below threshold."""
        return max(0.0, self.cfg.tau_h1 - p_h1)

    def step(
        self,
        p_h1: float,
        current_kappa: float,
        dt: float,
    ) -> float:
        """Execute one PI step.

        Returns
        -------
        kappa_new : updated coupling gain.
        """
        error = self.compute_error(p_h1)

        self._integral += error * dt
        self._integral = np.clip(
            self._integral,
            self.cfg.integral_min,
            self.cfg.integral_max,
        )

        delta_kappa = self.cfg.Kp * error + self.cfg.Ki * self._integral
        kappa_new = current_kappa + delta_kappa
        kappa_new = max(self.cfg.kappa_min, min(self.cfg.kappa_max, kappa_new))

        self._last_error = error
        self._p_h1_history.append(p_h1)
        self._kappa_history.append(kappa_new)
        self._error_history.append(error)
        if len(self._p_h1_history) > self.cfg.history_len:
            self._p_h1_history = self._p_h1_history[-self.cfg.history_len :]
            self._kappa_history = self._kappa_history[-self.cfg.history_len :]
            self._error_history = self._error_history[-self.cfg.history_len :]

        return kappa_new

    def is_gate_open(self, p_h1: float | None = None) -> bool:
        """Check if p_h1 > τ_h1 (consciousness gate open)."""
        val = (
            p_h1
            if p_h1 is not None
            else (self._p_h1_history[-1] if self._p_h1_history else 0.0)
        )
        return val > self.cfg.tau_h1

    def get_state(self) -> dict:
        """Return serialisable controller state."""
        return {
            "last_error": round(self._last_error, 6),
            "integral": round(self._integral, 6),
            "tau_h1": self.cfg.tau_h1,
            "gate_open": self.is_gate_open(),
            "p_h1_mean": round(
                float(np.mean(self._p_h1_history)) if self._p_h1_history else 0.0,
                6,
            ),
            "kappa_mean": round(
                float(np.mean(self._kappa_history)) if self._kappa_history else 0.0,
                6,
            ),
        }

    def reset(self) -> None:
        """Reset controller state."""
        self._integral = 0.0
        self._last_error = 0.0
        self._p_h1_history.clear()
        self._kappa_history.clear()
        self._error_history.clear()

File: backfire-kernel/crates/backfire-observers/src/tcbo.rs
// ─────────────────────────────────────────────────────────────────────
// Director-Class AI — TCBO (Topological Boundary Observable)
// ─────────────────────────────────────────────────────────────────────
//! Topological Boundary Observable (TCBO).
//!
//! Extracts a single scalar p_h1(t) in [0, 1] from multichannel phase data
//! via persistent homology (H1 cycles). When p_h1 > tau_h1 (default 0.72),
//! the boundary gate opens.
//!
//! Pipeline:
//!   1. Multichannel signal -> delay embedding (Takens' theorem)
//!   2. Sliding window -> point cloud
//!   3. H1 persistence estimation (spread-based fallback)
//!   4. Max H1 persistence -> logistic squash -> p_h1
//!
//! Note: This Rust implementation uses a spread-based fallback for persistence
//! estimation. For production TDA, use the Python TCBO with ripser.

use serde::{Deserialize, Serialize};

/// TCBO observer configuration.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TCBOConfig {
    pub embed_dim: usize,
    pub tau_delay: usize,
    pub window_size: usize,
    pub tau_h1: f64,
    pub beta: f64,
    pub s0: f64,
    pub persistence_threshold: f64,
    pub subsample_max: usize,
    pub compute_every_n: u64,
}

impl Default for TCBOConfig {
    fn default() -> Self {
        Self {
            embed_dim: 3,
            tau_delay: 1,
            window_size: 50,
            tau_h1: 0.72,
            beta: 8.0,
            s0: 0.0,
            persistence_threshold: 0.05,
            subsample_max: 500,
            compute_every_n: 1,
        }
    }
}

/// TCBO observer state snapshot.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TCBOState {
    pub p_h1: f64,
    pub s_h1: f64,
    pub is_conscious: bool,
    pub h1_error: f64,
    pub tau_h1: f64,
    pub buffer_size: usize,
}

/// Logistic squashing: p_h1 = 1 / (1 + exp(-beta * (s_h1 - s0))).
pub fn persistence_to_probability(s_h1: f64, s0: f64, beta: f64) -> f64 {
    let arg = (-beta * (s_h1 - s0)).clamp(-500.0, 500.0);
    1.0 / (1.0 + arg.exp())
}

/// Compute s0 so that persistence_to_probability(s_tau, s0, beta) == p.
pub fn s0_for_threshold(s_tau: f64, p: f64, beta: f64) -> f64 {
    let p_clamped = p.clamp(1e-12, 1.0 - 1e-12);
    s_tau - (1.0 / beta) * (p_clamped / (1.0 - p_clamped)).ln()
}

/// Delay-coordinate embedding of a 1-D signal.
///
/// Returns a matrix of shape (T_out, embed_dim) where T_out = T - (m-1)*tau.
pub fn delay_embed(x: &[f64], embed_dim: usize, tau_delay: usize) -> Vec<Vec<f64>> {
    let t = x.len();
    let offset = (embed_dim - 1) * tau_delay;
    if offset >= t {
        return Vec::new();
    }
    let out_t = t - offset;
    let mut z = vec![vec![0.0; embed_dim]; out_t];
    for k in 0..embed_dim {
        let start = offset - k * tau_delay;
        for i in 0..out_t {
            z[i][k] = x[start + i];
        }
    }
    z
}

/// Delay-embed each channel then concatenate columns.
///
/// Input: rows x n_channels. Output: rows x (n_channels * embed_dim).
pub fn delay_embed_multi(
    data: &[Vec<f64>],
    n_channels: usize,
    embed_dim: usize,
    tau_delay: usize,
) -> Vec<Vec<f64>> {
    if data.is_empty() || n_channels == 0 {
        return Vec::new();
    }

    // Extract each channel as a column
    let mut channels: Vec<Vec<f64>> = Vec::with_capacity(n_channels);
    for ch in 0..n_channels {
        let col: Vec<f64> = data.iter().map(|row| {
            if ch < row.len() { row[ch] } else { 0.0 }
        }).collect();
        channels.push(col);
    }

    // Delay-embed each channel
    let mut parts: Vec<Vec<Vec<f64>>> = Vec::with_capacity(n_channels);
    for col in &channels {
        let embedded = delay_embed(col, embed_dim, tau_delay);
        if embedded.is_empty() {
            return Vec::new();
        }
        parts.push(embedded);
    }

    // Horizontally concatenate
    let out_t = parts[0].len();
    let total_cols = n_channels * embed_dim;
    let mut result = vec![vec![0.0; total_cols]; out_t];
    for (ch_idx, part) in parts.iter().enumerate() {
        let col_offset = ch_idx * embed_dim;
        for i in 0..out_t {
            for k in 0..embed_dim {
                result[i][col_offset + k] = part[i][k];
            }
        }
    }
    result
}

/// Fallback H1 persistence estimator (no ripser dependency).
///
/// Estimates H1 persistence from point-cloud spread via radial std deviation.
fn persistence_fallback(cloud: &[Vec<f64>]) -> f64 {
    if cloud.is_empty() {
        return 0.0;
    }
    let d = cloud[0].len();
    let n = cloud.len() as f64;

    // Compute centroid
    let mut centroid = vec![0.0; d];
    for row in cloud {
        for (j, &v) in row.iter().enumerate() {
            centroid[j] += v;
        }
    }
    for c in centroid.iter_mut() {
        *c /= n;
    }

    // Compute radial distances from centroid
    let mut radials = Vec::with_capacity(cloud.len());
    for row in cloud {
        let r2: f64 = row.iter().zip(centroid.iter()).map(|(&v, &c)| (v - c).powi(2)).sum();
        radials.push(r2.sqrt());
    }

    // Std of radial distances
    let mean_r = radials.iter().sum::<f64>() / n;
    let var_r = radials.iter().map(|&r| (r - mean_r).powi(2)).sum::<f64>() / n;
    let spread = var_r.sqrt();

    if spread > 1e-12 {
        // Synthetic H1 persistence: birth=0.15*spread, death=0.85*spread
        0.85 * spread - 0.15 * spread // = 0.70 * spread
    } else {
        0.0
    }
}

/// Topological Boundary Observable.
///
/// Produces p_h1(t) in [0, 1] from multichannel phase data.
pub struct TCBOObserver {
    n: usize,
    cfg: TCBOConfig,
    buffer: Vec<Vec<f64>>,
    max_buffer: usize,
    pub p_h1: f64,
    pub s_h1: f64,
    pub is_conscious: bool,
    pub h1_error: f64,
    step_count: u64,
}

impl TCBOObserver {
    pub fn new(n: usize, config: TCBOConfig) -> Self {
        let max_buffer = config.window_size + (config.embed_dim - 1) * config.tau_delay + 10;
        let h1_error = config.tau_h1;
        Self {
            n,
            cfg: config,
            buffer: Vec::new(),
            max_buffer,
            p_h1: 0.0,
            s_h1: 0.0,
            is_conscious: false,
            h1_error,
            step_count: 0,
        }
    }

    pub fn default_params(n: usize) -> Self {
        Self::new(n, TCBOConfig::default())
    }

    /// Push a new phase vector into the rolling buffer.
    pub fn push(&mut self, theta: &[f64]) {
        self.buffer.push(theta.to_vec());
        if self.buffer.len() > self.max_buffer {
            let start = self.buffer.len() - self.max_buffer;
            self.buffer = self.buffer[start..].to_vec();
        }
    }

    /// Compute p_h1 from the current buffer.
    pub fn compute(&mut self, force: bool) -> f64 {
        self.step_count += 1;
        if !force && self.cfg.compute_every_n > 0 && (self.step_count % self.cfg.compute_every_n != 0) {
            return self.p_h1;
        }

        let min_needed = (self.cfg.embed_dim - 1) * self.cfg.tau_delay + self.cfg.window_size;
        if self.buffer.len() < min_needed {
            return self.p_h1;
        }

        let start = self.buffer.len() - min_needed;
        let signal = &self.buffer[start..];

        let z = delay_embed_multi(signal, self.n, self.cfg.embed_dim, self.cfg.tau_delay);
        if z.is_empty() {
            return self.p_h1;
        }

        // Take the last window_size rows as the point cloud
        let cloud_start = z.len().saturating_sub(self.cfg.window_size);
        let cloud = &z[cloud_start..];

        // Compute H1 persistence via fallback
        let lifetime = persistence_fallback(cloud);

        if lifetime > self.cfg.persistence_threshold {
            self.s_h1 = lifetime;
        } else {
            self.s_h1 = 0.0;
        }

        self.p_h1 = persistence_to_probability(self.s_h1, self.cfg.s0, self.cfg.beta);
        self.is_conscious = self.p_h1 > self.cfg.tau_h1;
        self.h1_error = (self.cfg.tau_h1 - self.p_h1).max(0.0);

        self.p_h1
    }

    /// Push + compute in one call.
    pub fn push_and_compute(&mut self, theta: &[f64], force: bool) -> f64 {
        self.push(theta);
        self.compute(force)
    }

    /// Return serialisable observer state.
    pub fn get_state(&self) -> TCBOState {
        TCBOState {
            p_h1: self.p_h1,
            s_h1: self.s_h1,
            is_conscious: self.is_conscious,
            h1_error: self.h1_error,
            tau_h1: self.cfg.tau_h1,
            buffer_size: self.buffer.len(),
        }
    }

    /// Clear buffer and reset state.
    pub fn reset(&mut self) {
        self.buffer.clear();
        self.p_h1 = 0.0;
        self.s_h1 = 0.0;
        self.is_conscious = false;
        self.h1_error = self.cfg.tau_h1;
        self.step_count = 0;
    }
}

/// TCBO PI controller configuration.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TCBOControllerConfig {
    pub tau_h1: f64,
    pub kp: f64,
    pub ki: f64,
    pub integral_min: f64,
    pub integral_max: f64,
    pub kappa_min: f64,
    pub kappa_max: f64,
    pub history_len: usize,
}

impl Default for TCBOControllerConfig {
    fn default() -> Self {
        Self {
            tau_h1: 0.72,
            kp: 0.8,
            ki: 0.2,
            integral_min: -5.0,
            integral_max: 5.0,
            kappa_min: 0.0,
            kappa_max: 5.0,
            history_len: 100,
        }
    }
}

/// PI controller driving gap-junction kappa from p_h1 deficit.
///
/// Error signal: e_h1(t) = max(0, tau_h1 - p_h1(t))  (deficit-only)
pub struct TCBOController {
    cfg: TCBOControllerConfig,
    integral: f64,
    last_error: f64,
    p_h1_history: Vec<f64>,
    kappa_history: Vec<f64>,
    error_history: Vec<f64>,
}

impl TCBOController {
    pub fn new(config: TCBOControllerConfig) -> Self {
        Self {
            cfg: config,
            integral: 0.0,
            last_error: 0.0,
            p_h1_history: Vec::new(),
            kappa_history: Vec::new(),
            error_history: Vec::new(),
        }
    }

    pub fn default_params() -> Self {
        Self::new(TCBOControllerConfig::default())
    }

    /// Deficit-only error: positive when below threshold.
    pub fn compute_error(&self, p_h1: f64) -> f64 {
        (self.cfg.tau_h1 - p_h1).max(0.0)
    }

    /// Execute one PI step. Returns kappa_new.
    pub fn step(&mut self, p_h1: f64, current_kappa: f64, dt: f64) -> f64 {
        let error = self.compute_error(p_h1);

        self.integral += error * dt;
        self.integral = self.integral.clamp(self.cfg.integral_min, self.cfg.integral_max);

        let delta_kappa = self.cfg.kp * error + self.cfg.ki * self.integral;
        let kappa_new = (current_kappa + delta_kappa).clamp(self.cfg.kappa_min, self.cfg.kappa_max);

        self.last_error = error;
        self.p_h1_history.push(p_h1);
        self.kappa_history.push(kappa_new);
        self.error_history.push(error);

        if self.p_h1_history.len() > self.cfg.history_len {
            let start = self.p_h1_history.len() - self.cfg.history_len;
            self.p_h1_history = self.p_h1_history[start..].to_vec();
            self.kappa_history = self.kappa_history[start..].to_vec();
            self.error_history = self.error_history[start..].to_vec();
        }

        kappa_new
    }

    /// Check if p_h1 > tau_h1 (boundary gate open).
    pub fn is_gate_open(&self, p_h1: Option<f64>) -> bool {
        let val = p_h1.unwrap_or_else(|| {
            self.p_h1_history.last().copied().unwrap_or(0.0)
        });
        val > self.cfg.tau_h1
    }

    /// Reset controller state.
    pub fn reset(&mut self) {
        self.integral = 0.0;
        self.last_error = 0.0;
        self.p_h1_history.clear();
        self.kappa_history.clear();
        self.error_history.clear();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_persistence_to_probability_midpoint() {
        let p = persistence_to_probability(0.0, 0.0, 8.0);
        assert!((p - 0.5).abs() < 1e-9, "p(0, 0, 8) should be 0.5, got {p}");
    }

    #[test]
    fn test_persistence_to_probability_high() {
        let p = persistence_to_probability(1.0, 0.0, 8.0);
        assert!(p > 0.99, "p(1, 0, 8) should be >0.99, got {p}");
    }

    #[test]
    fn test_persistence_to_probability_low() {
        let p = persistence_to_probability(-1.0, 0.0, 8.0);
        assert!(p < 0.01, "p(-1, 0, 8) should be <0.01, got {p}");
    }

    #[test]
    fn test_s0_for_threshold_roundtrip() {
        let s_tau = 0.5;
        let p_target = 0.72;
        let beta = 8.0;
        let s0 = s0_for_threshold(s_tau, p_target, beta);
        let p = persistence_to_probability(s_tau, s0, beta);
        assert!((p - p_target).abs() < 1e-6, "Roundtrip failed: got {p}");
    }

    #[test]
    fn test_delay_embed_basic() {
        let x: Vec<f64> = (0..10).map(|i| i as f64).collect();
        let z = delay_embed(&x, 3, 1);
        assert_eq!(z.len(), 8, "Should have 8 rows for T=10, m=3, tau=1");
        assert_eq!(z[0].len(), 3);
        // First row: x[2], x[1], x[0] (offset=2, k=0→start=2, k=1→start=1, k=2→start=0)
        assert!((z[0][0] - 2.0).abs() < 1e-9);
        assert!((z[0][1] - 1.0).abs() < 1e-9);
        assert!((z[0][2] - 0.0).abs() < 1e-9);
    }

    #[test]
    fn test_delay_embed_too_short() {
        let x = vec![1.0, 2.0];
        let z = delay_embed(&x, 3, 2);
        assert!(z.is_empty(), "Signal too short for embedding");
    }

    #[test]
    fn test_delay_embed_multi_basic() {
        let data: Vec<Vec<f64>> = (0..10).map(|i| vec![i as f64, (i as f64) * 2.0]).collect();
        let z = delay_embed_multi(&data, 2, 3, 1);
        assert_eq!(z.len(), 8);
        assert_eq!(z[0].len(), 6); // 2 channels * 3 embed_dim
    }

    #[test]
    fn test_persistence_fallback_spread() {
        // Cloud with some spread should give non-zero persistence
        let cloud: Vec<Vec<f64>> = (0..50)
            .map(|i| {
                let angle = i as f64 * 0.1;
                vec![angle.cos(), angle.sin(), angle * 0.1]
            })
            .collect();
        let p = persistence_fallback(&cloud);
        assert!(p > 0.0, "Spread cloud should have positive persistence, got {p}");
    }

    #[test]
    fn test_persistence_fallback_zero_cloud() {
        let cloud: Vec<Vec<f64>> = (0..10).map(|_| vec![1.0, 2.0, 3.0]).collect();
        let p = persistence_fallback(&cloud);
        assert!(p.abs() < 1e-9, "Constant cloud should have ~0 persistence, got {p}");
    }

    #[test]
    fn test_tcbo_observer_initial_state() {
        let obs = TCBOObserver::default_params(16);
        let state = obs.get_state();
        assert!((state.p_h1).abs() < 1e-9);
        assert!(!state.is_conscious);
        assert_eq!(state.buffer_size, 0);
    }

    #[test]
    fn test_tcbo_observer_push_and_compute() {
        let mut obs = TCBOObserver::default_params(16);
        // Push enough diverse data to trigger computation
        for i in 0..60 {
            let theta: Vec<f64> = (0..16)
                .map(|j| (i as f64 * 0.1 + j as f64 * 0.5).sin())
                .collect();
            obs.push(&theta);
        }
        let p = obs.compute(true);
        assert!(p >= 0.0 && p <= 1.0, "p_h1 should be in [0,1], got {p}");
    }

    #[test]
    fn test_tcbo_observer_reset() {
        let mut obs = TCBOObserver::default_params(16);
        for i in 0..60 {
            let theta: Vec<f64> = (0..16).map(|j| (i + j) as f64 * 0.1).collect();
            obs.push(&theta);
        }
        obs.compute(true);
        obs.reset();
        let state = obs.get_state();
        assert!((state.p_h1).abs() < 1e-9);
        assert_eq!(state.buffer_size, 0);
    }

    #[test]
    fn test_tcbo_controller_deficit_error() {
        let ctrl = TCBOController::default_params();
        let e = ctrl.compute_error(0.5);
        assert!((e - 0.22).abs() < 1e-9, "Error should be 0.72-0.5=0.22, got {e}");
    }

    #[test]
    fn test_tcbo_controller_no_error_above_threshold() {
        let ctrl = TCBOController::default_params();
        let e = ctrl.compute_error(0.9);
        assert!(e.abs() < 1e-9, "Error should be 0 above threshold, got {e}");
    }

    #[test]
    fn test_tcbo_controller_step_increases_kappa() {
        let mut ctrl = TCBOController::default_params();
        let kappa = ctrl.step(0.3, 1.0, 0.01);
        assert!(kappa > 1.0, "Kappa should increase when below threshold, got {kappa}");
    }

    #[test]
    fn test_tcbo_controller_gate() {
        let ctrl = TCBOController::default_params();
        assert!(!ctrl.is_gate_open(Some(0.5)));
        assert!(ctrl.is_gate_open(Some(0.8)));
    }

    #[test]
    fn test_tcbo_controller_kappa_clamped() {
        let mut ctrl = TCBOController::default_params();
        // Drive hard to saturate
        let mut kappa = 0.0;
        for _ in 0..1000 {
            kappa = ctrl.step(0.0, kappa, 1.0);
        }
        assert!(kappa <= ctrl.cfg.kappa_max, "Kappa {kappa} exceeds max");
    }

    #[test]
    fn test_tcbo_controller_reset() {
        let mut ctrl = TCBOController::default_params();
        ctrl.step(0.5, 1.0, 0.01);
        ctrl.reset();
        assert!(ctrl.last_error.abs() < 1e-9);
        assert!(ctrl.integral.abs() < 1e-9);
    }
}
